### PR TITLE
Users DbContext and FolderStructure Created

### PR DESCRIPTION
**Pull request by ChiragMajethiyaTatvasoft**

### CODE REVIEW

File: CIR.Application/Services/Users/UserService.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		public async Task<IActionResult> GetUserById(int id)
+		{
+			return await userRepository.GetUserById(id);
+		}
+		public async Task<Boolean> UserExists(string email, long id)
+		{
+			return await userRepository.UserExists(email, id);
+		}
+		public Task<IActionResult> CreateOrUpdateUser(User user)
+		{
```
Comment (GohilAmit) [inline CIR.Application/Services/Users/UserService.cs:23]: use async and await

File: CIR.Application/Services/Users/UserService.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+			return userRepository.CreateOrUpdateUser(user);
+		}
+		public async Task<IActionResult> DeleteUser(int id)
+		{
+			return await userRepository.DeleteUser(id);
+		}
+		public Task<IActionResult> GetAllUsersDetailBySP(int displayLength, int displayStart, string? sortCol, string? sortDir, string? search, int roleId, bool? enabled = null)
+		{
```
Comment (GohilAmit) [inline CIR.Application/Services/Users/UserService.cs:31]: use async and await

File: CIR.Common/Helper/EmailGeneration.cs [hunk 3]
```diff
@@ -6,61 +6,68 @@
...
 {
-    public class EmailGeneration
-    {
-        private readonly EmailModel _emailModel;
-        public EmailGeneration(IOptions<EmailModel> emailModel)
-        {
-            _emailModel = emailModel.Value;
+	public class EmailGeneration
+	{
+		private readonly EmailModel _emailModel;
```
Comment (GohilAmit) [inline CIR.Common/Helper/EmailGeneration.cs:11]: not use underscore

File: CIR.Data/Data/Users/UserRepository.cs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	{
+		#region Properties
+		private readonly CIRDbContext cIRDbContext;
+		private readonly EmailGeneration emailGeneration;
+		#endregion
+
+		#region Constructor
+		public UserRepository(CIRDbContext context, EmailGeneration emailService)
+		{
+			cIRDbContext = context;
```
Comment (GohilAmit) [inline CIR.Data/Data/Users/UserRepository.cs:24]: please define argument null exception like this in constructor :
cIRDbContext = context ??
throw new ArgumentNullException(nameof(context));

File: CIR.Data/Data/Users/UserRepository.cs [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+		}
+
+		/// <summary>
+		/// this meethod checks if user exists or not based on input user email
+		/// </summary>
+		/// <param name="email"></param>
+		/// <param name="id"></param>
+		/// <returns> if user exists true else false </returns>
+		public async Task<Boolean> UserExists(string email, long id)
+		{
```
Comment (GohilAmit) [inline CIR.Data/Data/Users/UserRepository.cs:68]: Use StoreProcedure remove linq query

File: CIR.Data/Data/Users/UserRepository.cs [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+		/// </summary>
+		/// <param name="id"></param>
+		/// <returns> deleted/disabled user </returns>
+		public async Task<IActionResult> DeleteUser(int id)
+		{
+			try
+			{
+				if (id == 0)
+				{
+					return new JsonResult(new CustomResponse<string>() { StatusCode = (int)HttpStatusCodesAndMessages.HttpStatus.NotFound, Result = false, Message = HttpStatusCodesAndMessages.HttpStatus.NotFound.GetDescriptionAttribute(), Data = SystemMessages.msgInvalidId });
```
Comment (GohilAmit) [inline CIR.Data/Data/Users/UserRepository.cs:151]: need to changes return response

File: CIR/Controllers/Users/UsersController.cs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+namespace CIR.Controllers.Users
+{
+	[Route("api/[controller]")]
+	[ApiController]
+	public class UsersController : ControllerBase
+	{
+		#region Properties
+		private readonly IUserService userService;
+		private readonly ILogger<UsersController> loggerr;
```
Comment (GohilAmit) [inline CIR/Controllers/Users/UsersController.cs:15]: remove ILogger

File: CIR/Controllers/Users/UsersController.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+		[HttpGet("{id}")]
+		public async Task<IActionResult> GetUserById(int id)
+		{
+			try
+			{
+				return await userService.GetUserById(id);
+			}
+			catch (Exception ex)
+			{
+				return new JsonResult(new CustomResponse<Exception>() { StatusCode = (int)HttpStatusCodesAndMessages.HttpStatus.InternalServerError, Result = false, Message = HttpStatusCodesAndMessages.HttpStatus.InternalServerError.GetDescriptionAttribute(), Data = ex });
```
Comment (GohilAmit) [inline CIR/Controllers/Users/UsersController.cs:42]: need to change catch return response

File: CIR/Controllers/Users/UsersController.cs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+					else
+					{
+						return await userService.CreateOrUpdateUser(user);
+
+					}
+
+				}
+				catch (Exception ex)
+				{
+					return new JsonResult(new CustomResponse<Exception>() { StatusCode = (int)HttpStatusCodesAndMessages.HttpStatus.InternalServerError, Result = false, Message = HttpStatusCodesAndMessages.HttpStatus.InternalServerError.GetDescriptionAttribute(), Data = ex });
```
Comment (GohilAmit) [inline CIR/Controllers/Users/UsersController.cs:72]: need to change catch return response

File: CIR/Controllers/Users/UsersController.cs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+					{
+						return new JsonResult(new CustomResponse<string>() { StatusCode = (int)HttpStatusCodesAndMessages.HttpStatus.BadRequest, Result = false, Message = SystemMessages.msgBadRequest });
+					}
+					else
+					{
+						return await userService.CreateOrUpdateUser(user);
+					}
+				}
+				catch (Exception ex)
+				{
```
Comment (GohilAmit) [inline CIR/Controllers/Users/UsersController.cs:101]: need to change catch return response

File: CIR/Controllers/Users/UsersController.cs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+		public async Task<IActionResult> Delete(int id)
+		{
+			if (ModelState.IsValid)
+			{
+				try
+				{
+					return await userService.DeleteUser(id);
+				}
+				catch (Exception ex)
+				{
```
Comment (GohilAmit) [inline CIR/Controllers/Users/UsersController.cs:122]: need to change catch return response

File: CIR/Controllers/Users/UsersController.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+		public async Task<IActionResult> GetAllUsersDetailBySP(int displayLength, int displayStart, string? sortCol, string? search, string? sortDir, int roleId, bool? enabled = null)
+		{
+			try
+			{
+				search ??= string.Empty;
+				return await userService.GetAllUsersDetailBySP(displayLength, displayStart, sortCol, search, sortDir, roleId, enabled);
+			}
+			catch (Exception ex)
+			{
+				return new JsonResult(new CustomResponse<Exception>() { StatusCode = (int)HttpStatusCodesAndMessages.HttpStatus.InternalServerError, Result = false, Message = HttpStatusCodesAndMessages.HttpStatus.InternalServerError.GetDescriptionAttribute(), Data = ex });
```
Comment (GohilAmit) [inline CIR/Controllers/Users/UsersController.cs:150]: need to change catch return response

File: CIR.Common/Helper/EmailGeneration.cs (hunk 3 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR.Common/Helper/EmailGeneration.cs]: Done

File: CIR.Data/Data/Users/UserRepository.cs (hunk 4 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR.Data/Data/Users/UserRepository.cs]: Done

File: CIR.Data/Data/Users/UserRepository.cs (hunk 5 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR.Data/Data/Users/UserRepository.cs:69]: Done

File: CIR.Data/Data/Users/UserRepository.cs (hunk 6 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR.Data/Data/Users/UserRepository.cs]: Done

File: CIR/Controllers/Users/UsersController.cs (hunk 7 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR/Controllers/Users/UsersController.cs:15]: Done

File: CIR/Controllers/Users/UsersController.cs (hunk 8 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR/Controllers/Users/UsersController.cs]: Done

File: CIR/Controllers/Users/UsersController.cs (hunk 9 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR/Controllers/Users/UsersController.cs]: Done

File: CIR/Controllers/Users/UsersController.cs (hunk 10 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR/Controllers/Users/UsersController.cs:100]: Done

File: CIR/Controllers/Users/UsersController.cs (hunk 11 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR/Controllers/Users/UsersController.cs:121]: Done

File: CIR/Controllers/Users/UsersController.cs (hunk 12 above)
Comment (ChiragMajethiyaTatvasoft) [reply CIR/Controllers/Users/UsersController.cs]: Done

**Review by GohilAmit — APPROVED** [review]

**Review by jayk-patel-tatvasoft — APPROVED** [review]

**Pull request merged**